### PR TITLE
[FLINK-35242] Supports per-SE type configuration & "lenient" evolution behavior

### PR DESCRIPTION
**Pull request by yuxiqian**
This closes FLINK-35242.

* Adds `TRY_EVOLVE` behavior that tolerates exception during metadata applying process
* Supports configuring various behavior for each schema evolution type with `include.schema.changes` and `exclude.schema.changes` sink option
* Adds schema evolution IT cases

---

With this change, Schema Operator now stores both "upstream" schema (which always keep up with the structure coming from pipeline source **after the transformation**) and "evolved" schema (the schema that actually applied to sink). They might differ when:

* Schema change events are partially or fully ignored (by fine-grained options / `IGNORE` behavior)
* Some schema change events failed to apply but pipeline keeps executing (in `TRY_EVOLVE` mode)

If these cases occur, `SchemaOperator` needs to cast upstream schema to match the actual evolved version (by adding / removing columns, converting types, etc.) with tolerance (fill in `null` if meaningful casts are not possible.)

### CODE REVIEW

File: flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/translator/SchemaOperatorTranslator.java [hunk 1]
```diff
@@ -58,12 +59,18 @@ public DataStream<Event> translate(
...
+                return dropSchemaChangeEvent(
+                        addSchemaOperator(input, parallelism, metadataApplier, routes, true),
+                        parallelism);
+            case TRY_EVOLVE:
+                return addSchemaOperator(input, parallelism, metadataApplier, routes, true);
+            case EVOLVE:
+                return addSchemaOperator(input, parallelism, metadataApplier, routes, false);
             case EXCEPTION:
-                return exceptionOnSchemaChange(input, parallelism);
+                return exceptionOnSchemaChange(
```
[the remark this answers is not in the document]
Comment (yuxiqian) [reply flink-cdc-composer/src/main/java/org/apache/flink/cdc/composer/flink/translator/SchemaOperatorTranslator.java:71]: Thanks for pointing it out! I think it would be better fix it in this PR since SchemaOperator & Registry will be greatly modified to implement FLINK-35242.

(Seems #3352 / FLINK-35432 is about an irrelevant problem about MySQL ddl parsing, and should be fine to be reviewed and merged.)

**Comment by yuxiqian** [comment]
@PatrickRen @leonardBang PTAL

**Comment by yuxiqian** [comment]
Just added `LENIENT` schema evolve behavior & rebased to `master`. Could @loserwang1024 please take a look at this?

Since this PR has grown too big, hopefully this could be reviewed & merged sooner to avoid any conflicts hassle in the future. cc @leonardBang

**Comment by leonardBang** [comment]
@yuxiqian Thanks for the great work, could you rebase to latest master?

**Comment by yuxiqian** [comment]
Done, rebased with master branch

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/sink/DataSink.java [hunk 2]
```diff
@@ -30,5 +33,5 @@ public interface DataSink {
     EventSinkProvider getEventSinkProvider();
 
     /** Get the {@link MetadataApplier} for applying metadata changes to external systems. */
-    MetadataApplier getMetadataApplier();
+    MetadataApplier getMetadataApplier(Set<SchemaChangeEventType> enabledEventTypes);
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/sink/DataSink.java:36]: This is a public API, we should keep the backward compatibility

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/sink/MetadataApplier.java [hunk 3]
```diff
@@ -19,13 +19,22 @@
...
 /** {@code MetadataApplier} is used to apply metadata changes to external systems. */
 @PublicEvolving
 public interface MetadataApplier extends Serializable {
 
+    /** Checks if this metadata applier should handle this event type. */
+    boolean acceptsSchemaEvolutionType(SchemaChangeEventType schemaChangeEventType);
+
+    /** Checks what kind of schema change events downstream can handle. */
+    Set<SchemaChangeEventType> getSupportedSchemaEvolutionTypes();
+
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/sink/MetadataApplier.java:37]: Please consider the compatibility when change existing public API. Imaging the case that user has a custom pipeline connector, is it still work after he/she bump the flink cdc version?

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/event/SchemaChangeEventType.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        } else if (event instanceof AlterColumnTypeEvent) {
+            return ALTER_COLUMN_TYPE;
+        } else if (event instanceof CreateTableEvent) {
+            return CREATE_TABLE;
+        } else if (event instanceof DropColumnEvent) {
+            return DROP_COLUMN;
+        } else if (event instanceof RenameColumnEvent) {
+            return RENAME_COLUMN;
+        } else {
+            throw new RuntimeException("Unknown schema change event type: " + event.getClass());
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/event/SchemaChangeEventType.java:40]: Why a RuntimeException thrown here but `ofTag()` method return a `null` object?

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/event/SchemaChangeEventType.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.cdc.common.event;
+
+/** An enumeration of schema change event types for {@link SchemaChangeEvent}. */
+public enum SchemaChangeEventType {
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/event/SchemaChangeEventType.java:21]: This should be a public interface from my point.

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/event/SchemaChangeEventTypeFamily.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    public static final SchemaChangeEventType[] ALTER = {SchemaChangeEventType.ALTER_COLUMN_TYPE};
+
+    public static final SchemaChangeEventType[] CREATE = {SchemaChangeEventType.CREATE_TABLE};
+
+    public static final SchemaChangeEventType[] DROP = {SchemaChangeEventType.DROP_COLUMN};
+
+    public static final SchemaChangeEventType[] RENAME = {SchemaChangeEventType.RENAME_COLUMN};
+
+    public static final SchemaChangeEventType[] TABLE = {SchemaChangeEventType.CREATE_TABLE};
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/event/SchemaChangeEventTypeFamily.java:36]: Could you share the basis of the categories ?  current hierarchy confuse me a bit.

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/exceptions/SchemaEvolveException.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+ */
+
+package org.apache.flink.cdc.common.exceptions;
+
+import org.apache.flink.cdc.common.event.SchemaChangeEvent;
+
+import javax.annotation.Nullable;
+
+/** An exception occurred during schema evolution. */
+public class SchemaEvolveException extends Exception {
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/exceptions/SchemaEvolveException.java:25]: extends `FlinkRuntimeException` or `CDCRuntimeException` ?

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/pipeline/SchemaChangeBehavior.java [hunk 8]
```diff
@@ -22,7 +22,9 @@
 /** Behavior for handling schema changes. */
 @PublicEvolving
 public enum SchemaChangeBehavior {
-    EVOLVE,
     IGNORE,
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/pipeline/SchemaChangeBehavior.java:25]: you can add a note about the order adjustment

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-doris/src/main/java/org/apache/flink/cdc/connectors/doris/sink/DorisDataSink.java [hunk 9]
```diff
@@ -74,7 +76,7 @@ public EventSinkProvider getEventSinkProvider() {
     }
 
     @Override
-    public MetadataApplier getMetadataApplier() {
-        return new DorisMetadataApplier(dorisOptions, configuration);
+    public MetadataApplier getMetadataApplier(Set<SchemaChangeEventType> enabledEventTypes) {
+        return new DorisMetadataApplier(dorisOptions, configuration, enabledEventTypes);
     }
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-doris/src/main/java/org/apache/flink/cdc/connectors/doris/sink/DorisDataSink.java:81]: You need change all connectors in one PR if you bring an incompatible change in public API, we should avoid this kind of change.

File: flink-cdc-runtime/src/test/resources/log4j2-test.properties [hunk 10]
```diff
@@ -15,7 +15,7 @@
 
 # Set root logger level to OFF to not flood build logs
 # set manually to INFO for debugging purposes
-rootLogger.level = OFF
+rootLogger.level = DEBUG
```
Comment (leonardBang) [inline flink-cdc-runtime/src/test/resources/log4j2-test.properties:18]: INFO ？

File: flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/schema/metrics/SchemaOperatorMetrics.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+import org.apache.flink.cdc.common.pipeline.SchemaChangeBehavior;
+import org.apache.flink.cdc.runtime.operators.schema.SchemaOperator;
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.metrics.MetricGroup;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/** A collection class for handling metrics in {@link SchemaOperator}. */
+public class SchemaOperatorMetrics {
```
Comment (leonardBang) [inline flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/schema/metrics/SchemaOperatorMetrics.java:29]: We can split the metric logic to another PR in next time, it's totally independent with SE behavior. But, I like the idea to report metrics for SchemaOperator.

**Review by leonardBang — COMMENTED** [review]
Thanks @yuxiqian for the great work, I finished the first round review and left some comments and I think you can start to address my comments.

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/sink/DataSink.java (hunk 2 above)
Comment (yuxiqian) [reply flink-cdc-common/src/main/java/org/apache/flink/cdc/common/sink/DataSink.java]: Done, moved `setAcceptedSchemaEvolutionTypes` to `MetadataApplier` interface, so `DataSink` could be left unchanged.

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/sink/MetadataApplier.java (hunk 3 above)
Comment (yuxiqian) [reply flink-cdc-common/src/main/java/org/apache/flink/cdc/common/sink/MetadataApplier.java]: I've tried to add reasonable `default` implementations for new methods to break less downstream code.

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/event/SchemaChangeEventType.java (hunk 4 above)
Comment (yuxiqian) [reply flink-cdc-common/src/main/java/org/apache/flink/cdc/common/event/SchemaChangeEventType.java:43]: I thought since `SchemaChangeEventType#ofEvent` receives a `SchemaChangeEvent` which must correspond to a valid event type, any unknown event type is unexpected and implies some coding mistake. However, `SchemaChangeEventType#ofTag` receives a String which might not be a valid type, so I didn't throw an exception for that case. Maybe `Optional<>` is better?

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-doris/src/main/java/org/apache/flink/cdc/connectors/doris/sink/DorisDataSink.java (hunk 9 above)
Comment (yuxiqian) [reply flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-doris/src/main/java/org/apache/flink/cdc/connectors/doris/sink/DorisDataSink.java]: Left `DataSink` interface unchanged, and adds `default` implementation in `MetadataApplier` interface.

File: flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/schema/metrics/SchemaOperatorMetrics.java (hunk 11 above)
Comment (yuxiqian) [reply flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/schema/metrics/SchemaOperatorMetrics.java:29]: Thanks for the reminder, I'll take notice on this next time.

**Comment by yuxiqian** [comment]
Squashed & Rebased with `master`.

Since this also changes `SchemaManager` serialization format, could @lvyanquan please take a look? (Serialization version isn't updated since it's expected that no version will be released between this PR and FLINK-34638.

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/exceptions/SchemaEvolveException.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+import org.apache.flink.cdc.common.event.SchemaChangeEvent;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import javax.annotation.Nullable;
+
+/** An exception occurred during schema evolution. */
+public class SchemaEvolveException extends FlinkRuntimeException {
+    private final SchemaChangeEvent applyingEvent;
+    private final String problem;
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/exceptions/SchemaEvolveException.java:28]: exceptionMessage ?

Comment (leonardBang) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/exceptions/SchemaEvolveException.java:28]: exceptionMessage ?

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/exceptions/SchemaEvolveException.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    private final @Nullable Throwable context;
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/exceptions/SchemaEvolveException.java:29]: the name context is strange name for a Throwable type

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/event/SchemaChangeEventType.java [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.cdc.common.event;
+
+import org.apache.flink.cdc.common.annotation.Public;
+
+/** An enumeration of schema change event types for {@link SchemaChangeEvent}. */
+@Public
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/event/SchemaChangeEventType.java:23]: Please use `PublicEvolving` for first introducing API

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/event/SchemaChangeEventType.java (hunk 4 above)
Comment (leonardBang) [reply flink-cdc-common/src/main/java/org/apache/flink/cdc/common/event/SchemaChangeEventType.java:43]: We can simply throw a RuntimeException for invalid String tag, a null SchemaChangeEventType is dangerous for caller.

File: flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/schema/event/ApplyEvolvedSchemaChangeResponse.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+package org.apache.flink.cdc.runtime.operators.schema.event;
+
+import org.apache.flink.cdc.runtime.operators.schema.SchemaOperator;
+import org.apache.flink.cdc.runtime.operators.schema.coordinator.SchemaRegistry;
+import org.apache.flink.runtime.operators.coordination.CoordinationResponse;
+
+/**
+ * The response from {@link SchemaOperator} to {@link SchemaRegistry} to request apply evolved
+ * schema changes.
```
Comment (leonardBang) [inline flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/schema/event/ApplyEvolvedSchemaChangeResponse.java:26]: the doc looks like incorrect ?

File: flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/schema/event/ApplyUpstreamSchemaChangeRequest.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+import org.apache.flink.cdc.runtime.operators.schema.coordinator.SchemaRegistry;
+import org.apache.flink.runtime.operators.coordination.CoordinationRequest;
+
+import java.util.Objects;
+
+/**
+ * The request from {@link SchemaOperator} to {@link SchemaRegistry} to request apply evolved schema
+ * changes.
+ */
+public class ApplyUpstreamSchemaChangeRequest implements CoordinationRequest {
```
Comment (leonardBang) [inline flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/schema/event/ApplyUpstreamSchemaChangeRequest.java:32]:
```suggestion
/**
 * The request from {@link SchemaOperator} to {@link SchemaRegistry} to request apply original schema
 * changes.
 */
public class ApplyOriginalSchemaChangeRequest implements CoordinationRequest {
```

File: flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/schema/event/ApplyEvolvedSchemaChangeRequest.java [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.flink.cdc.runtime.operators.schema.SchemaOperator;
+import org.apache.flink.cdc.runtime.operators.schema.coordinator.SchemaRegistry;
+import org.apache.flink.runtime.operators.coordination.CoordinationRequest;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * The request from {@link SchemaOperator} to {@link SchemaRegistry} to request apply evolved schema
+ * changes.
```
Comment (leonardBang) [inline flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/schema/event/ApplyEvolvedSchemaChangeRequest.java:31]:
```suggestion
 * The request from {@link SchemaOperator} to {@link SchemaRegistry} to request apply evolved schema
 * changes, the evolved schema changes come from original schema changes with different schema evolution strategy.
```

File: flink-cdc-e2e-tests/flink-cdc-pipeline-e2e-tests/src/test/java/org/apache/flink/cdc/pipeline/tests/SchemaEvolveE2eITCase.java [hunk 18]
```diff
@@ -0,0 +1,758 @@
...
+        waitUntilSpecificEvent(
+                String.format(
+                        "Ignored schema change DropColumnEvent{tableId=%s.members, droppedColumnNames=[biological_sex]} to table %s.members.",
+                        schemaEvolveDatabase.getDatabaseName(),
+                        schemaEvolveDatabase.getDatabaseName()),
+                jobManagerConsumer);
+    }
+
+    @Test
+    public void testLenientSchemaEvolution() throws Exception {
```
Comment (leonardBang) [inline flink-cdc-e2e-tests/flink-cdc-pipeline-e2e-tests/src/test/java/org/apache/flink/cdc/pipeline/tests/SchemaEvolveE2eITCase.java:657]: minor: I like the these tests, but parameterized tests will make the test code shorter

File: flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/schema/event/ApplyOriginalSchemaChangeRequest.java [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+import org.apache.flink.cdc.common.event.SchemaChangeEvent;
+import org.apache.flink.cdc.common.event.TableId;
+import org.apache.flink.cdc.runtime.operators.schema.SchemaOperator;
+import org.apache.flink.cdc.runtime.operators.schema.coordinator.SchemaRegistry;
+import org.apache.flink.runtime.operators.coordination.CoordinationRequest;
+
+import java.util.Objects;
+
+/**
+ * The request from {@link SchemaOperator} to {@link SchemaRegistry} to request apply evolved schema
```
Comment (leonardBang) [inline flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/schema/event/ApplyOriginalSchemaChangeRequest.java:29]: evolved -> original ?

**Review by leonardBang — APPROVED** [review]
Thanks @yuxiqian for the great work, LGTM
